pkg/keyspace/tso_keyspace_group.go

```diff
@@ -72,42 +122,137 @@ func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup,
 		err error
 	)
 
-	m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+	if err := m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
 		kg, err = m.store.LoadKeyspaceGroup(txn, id)
 		if err != nil {
 			return err
 		}
 		return nil
-	})
+	}); err != nil {
+		return nil, err
+	}
 	return kg, nil
 }
 
 // DeleteKeyspaceGroupByID deletes the keyspace group by id.
 func (m *GroupManager) DeleteKeyspaceGroupByID(id uint32) error {
-	m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+	var (
+		kg  *endpoint.KeyspaceGroup
+		err error
+	)
+
+	if err := m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+		kg, err = m.store.LoadKeyspaceGroup(txn, id)
+		if err != nil {
+			return err
+		}
 		return m.store.DeleteKeyspaceGroup(txn, id)
-	})
+	}); err != nil {
+		return err
+	}
+
+	userKind := endpoint.StringUserKind(kg.UserKind)
+	m.Lock()
+	defer m.Unlock()
+	// TODO: move out the keyspace to another group
+	// we don't need the keyspace group as the return value
+	m.groups[userKind].Remove(id)
+
 	return nil
 }
 
-func (m *GroupManager) saveKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup) error {
+func (m *GroupManager) saveKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup, overwrite bool) error {
 	return m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
 		for _, keyspaceGroup := range keyspaceGroups {
 			// TODO: add replica count
 			newKG := &endpoint.KeyspaceGroup{
-				ID:       keyspaceGroup.ID,
-				UserKind: keyspaceGroup.UserKind,
+				ID:        keyspaceGroup.ID,
+				UserKind:  keyspaceGroup.UserKind,
+				Keyspaces: keyspaceGroup.Keyspaces,
 			}
 			// Check if keyspace group has already existed.
 			oldKG, err := m.store.LoadKeyspaceGroup(txn, keyspaceGroup.ID)
 			if err != nil {
 				return err
 			}
-			if oldKG != nil {
+			if oldKG != nil && !overwrite {
 				return ErrKeyspaceGroupExists
 			}
 			m.store.SaveKeyspaceGroup(txn, newKG)
 		}
 		return nil
 	})
 }
+
+// GetAvailableKeyspaceGroupIDByKind returns the available keyspace group id by user kind.
+func (m *GroupManager) GetAvailableKeyspaceGroupIDByKind(userKind endpoint.UserKind) (string, error) {
+	m.RLock()
+	defer m.RUnlock()
+	groups, ok := m.groups[userKind]
+	if !ok {
+		return "", errors.Errorf("user kind %s not found", userKind)
+	}
+	kg := groups.Top()
+	return strconv.FormatUint(uint64(kg.ID), 10), nil
+}
+
+// UpdateKeyspaceForGroup updates the keyspace field for the keyspace group.
+func (m *GroupManager) UpdateKeyspaceForGroup(userKind endpoint.UserKind, groupID string, keyspaceID uint32, mutation string) error {
```

Comment:
For Bootstrap, is it necessary to batching the operations? Collects all keyspaces to add/remove for every keyspace group from preAlloc, then apply saveKeyspaceGroups for each group just once.

Reply:
slice.Remove() is an O(n), we might need a more efficient data structure in the future, considering we also need a more compact way to store ids, mentioned in another comment. We can change it in the separate PR.

Reply:
I just keep the original logic, which separates the preAlloc from the default keyspace. As for the second comment, I agree with it.